### PR TITLE
feat: [PL-41273]: Notification when connector goes down

### PR DESCRIPTION
**Pull request by jennyjames21**
## Describe your changes

Summary:
Add handler to send notification when connector goes down

Test
-Unit test
-Local sanity

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54535)

### CODE REVIEW

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

File: 440-connector-nextgen/src/main/java/io/harness/connector/entities/Connector.java [hunk 1]
```diff
@@ -105,6 +106,8 @@ public abstract class Connector implements PersistentEntity, NGAccountAccess, Gi
   @Getter(AccessLevel.NONE) @Setter(AccessLevel.NONE) Boolean isEntityInvalid;
   String invalidYamlString; // TODO: remove this field after RenameInvalidYamlStringToYamlMigration runs
   String yaml;
+  Long connectorDisconnectIteration;
+  Long lastAlertSent;
```
Comment (meenakshi-harness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/entities/Connector.java:110]: Should we move these inside connectivityDetails

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

File: 440-connector-nextgen/src/main/java/io/harness/connector/entities/Connector.java (hunk 1 above)
Comment (jennyjames21) [reply 440-connector-nextgen/src/main/java/io/harness/connector/entities/Connector.java]: yes makes sense. 
Moved   lastAlertSent inside connectivityDetails

keeping connectorDisconnectIteration inside the connector as it's handler job running on the entity.

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

File: 970-ng-commons/src/main/java/io/harness/constants/ConnectorConstants.java [hunk 2]
```diff
@@ -59,4 +59,5 @@ public class ConnectorConstants {
   public static final String CONNECTOR_TYPE_STATS = "Count of Connectors grouped by type.";
   public static final String CONNECTOR_STATUS_STATS = "Count of Connectors grouped by status.";
   public static final String CONNECTIVITY_MODE = "Connector connectivity mode on which the filter is applied";
+  public static final String LAST_ALERT = "Last alert sent time last when connector went down";
```
Comment (meenakshi-harness) [inline 970-ng-commons/src/main/java/io/harness/constants/ConnectorConstants.java:62]: can we get this reviewed by doc team?

File: 120-ng-manager/src/main/java/io/harness/ng/core/handler/connector/ConnectorDisconnectHandler.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  @Override
+  public void handle(Connector connector) {
+    if (isConnectorDisconnectAlertSend(connector)) {
+      return;
+    }
+    String notificationTriggerRequestId = generateUuid();
+    Map<String, String> templateData = new HashMap<>();
+    templateData.put("TEMPLATE_IDENTIFIER", "connector_down");
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/handler/connector/ConnectorDisconnectHandler.java:106]: use constants?

File: 120-ng-manager/src/main/java/io/harness/ng/core/handler/connector/ConnectorDisconnectHandler.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    NotificationTriggerRequest.Builder notificationTriggerRequestBuilder =
+        NotificationTriggerRequest.newBuilder()
+            .setId(notificationTriggerRequestId)
+            .setAccountId(connector.getAccountIdentifier())
+            .setOrgId(connector.getOrgIdentifier() != null ? connector.getOrgIdentifier() : "")
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/handler/connector/ConnectorDisconnectHandler.java:111]: why are we setting empty string and not null?

File: 120-ng-manager/src/main/java/io/harness/ng/core/handler/connector/ConnectorDisconnectHandler.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+          new Update().set(ConnectorKeys.alertSentAt, connector.getConnectivityDetails().getLastConnectedAt());
+      connectorRepository.update(connector.getAccountIdentifier(), connector.getOrgIdentifier(),
+          connector.getProjectIdentifier(), criteria, update);
+    }
+  }
+
+  private boolean isConnectorDisconnectAlertSend(Connector connector) {
+    return connector.getConnectivityDetails() != null && connector.getConnectivityDetails().getLastAlertSent() != null
+        && connector.getConnectivityDetails().getLastAlertSent()
+        == connector.getConnectivityDetails().getLastConnectedAt();
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/handler/connector/ConnectorDisconnectHandler.java:131]: store this in variable `connector.getConnectivityDetails()` and re-use

Comment (jennyjames21) [reply 120-ng-manager/src/main/java/io/harness/ng/core/handler/connector/ConnectorDisconnectHandler.java]: updated

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 4 of 6 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54535)* status: 6 file(s) to review. 2 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/handler/connector/ConnectorDisconnectHandler.java` line 113 at r7](https://reviewable.io/reviews/harness/harness-core/54535#-NibtxRu-D-XICvJAjoq:-NibtxRu-D-XICvJAjor:bxz2xo) ([raw file](https://github.com/harness/harness-core/blob/84aba577858463839706b7ab78aa45478b891962/120-ng-manager/src/main/java/io/harness/ng/core/handler/connector/ConnectorDisconnectHandler.java#L113)):*
> ```Java
>             .setAccountId(connector.getAccountIdentifier())
>             .setOrgId(connector.getOrgIdentifier() != null ? connector.getOrgIdentifier() : "")
>             .setProjectId(connector.getProjectIdentifier() != null ? connector.getProjectIdentifier() : "")
> ```

Same here. Why not null?

**Comment by jennyjames21** [comment]
trigger uts

**Review by meenakshi-harness — APPROVED** [review]

**Comment by jennyjames21** [comment]
trigger messagecheck

**Pull request merged**